controllers/binding/binding_adapter.go

```diff
@@ -0,0 +1,180 @@
+				return reconciler.RequeueOnErrorOrStop(err)
+			}
+		}
+	}
+
+	return reconciler.ContinueProcessing()
+}
+
+// createIntegrationPipelineRunWithEnvironment creates new integration PipelineRun. The Pipeline information and the parameters to it
+// will be extracted from the given integrationScenario. The integration's Snapshot will also be passed to the integration PipelineRun.
+// If the creation of the PipelineRun is unsuccessful, an error will be returned.
+func (a *Adapter) createIntegrationPipelineRunWithEnvironment(application *applicationapiv1alpha1.Application, integrationTestScenario *v1alpha1.IntegrationTestScenario, snapshot *applicationapiv1alpha1.Snapshot, environment *applicationapiv1alpha1.Environment) error {
+	deploymentTarget, err := a.GetDeploymentTargetForEnvironment(environment)
+	if err != nil || deploymentTarget == nil {
+		return err
+	}
+
+	pipelineRun := tekton.NewIntegrationPipelineRun(snapshot.Name, application.Namespace, *integrationTestScenario).
+		WithSnapshot(snapshot).
+		WithIntegrationLabels(integrationTestScenario).
+		WithEnvironmentAndDeploymentTarget(deploymentTarget, environment.Name).
+		AsPipelineRun()
+	// copy PipelineRun PAC annotations/labels from snapshot to integration test PipelineRuns
+	helpers.CopyAnnotationsByPrefix(&snapshot.ObjectMeta, &pipelineRun.ObjectMeta, gitops.PipelinesAsCodePrefix, gitops.PipelinesAsCodePrefix)
+	helpers.CopyLabelsByPrefix(&snapshot.ObjectMeta, &pipelineRun.ObjectMeta, gitops.PipelinesAsCodePrefix, gitops.PipelinesAsCodePrefix)
+	err = ctrl.SetControllerReference(snapshot, pipelineRun, a.client.Scheme())
+	if err != nil {
+		return err
+	}
+	err = a.client.Create(a.context, pipelineRun)
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+// GetDeploymentTargetForEnvironment gets the DeploymentTarget associated with Environdment, if the DeploymentTarget is not found, an error will be returned
+func (a *Adapter) GetDeploymentTargetForEnvironment(environment *applicationapiv1alpha1.Environment) (*applicationapiv1alpha1.DeploymentTarget, error) {
+	deploymentTargetClaim, err := a.GetDeploymentTargetClaimForEnvironment(environment)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find deploymentTargetClaim %s defined in environment %s: %w", deploymentTargetClaim.Name, environment.Name, err)
+	}
+
+	deploymentTarget, err := a.GetDeploymentTargetForDeploymentTargetClaim(deploymentTargetClaim)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find deploymentTarget %s: %w", deploymentTarget.Name, err)
+	}
+
+	return deploymentTarget, nil
+}
+
+// GetDeploymentTargetClaimForEnvironment try to find the DeploymentTargetClaim whose name is defined in Environment
+// if not found, an error is returned
+func (a *Adapter) GetDeploymentTargetClaimForEnvironment(environment *applicationapiv1alpha1.Environment) (*applicationapiv1alpha1.DeploymentTargetClaim, error) {
+	dtcName := environment.Spec.Configuration.Target.DeploymentTargetClaim.ClaimName
```

Comment:
Should we also check first if the `environment.Spec.Configuration.Target` defined to avoid errors here?

Reply:
Thanks, I concerned it and `dtc.Spec.TargetName`, so I tested them, the result is if the `Target` and `TargetName` are not defined, `environment.Spec.Configuration.Target.DeploymentTargetClaim.ClaimName` and `dtc.Spec.TargetName` will be "" and then will be caught by `if dtName == ""` and `if dtcName == ""`, therefore I didn't check `environment.Spec.Configuration.Target` first. But let me add code to handle it to make code review easier.

Reply:
Thanks!